Deploy a Micronaut application to AWS Lambda as an ARM64 native-image

I'd like to be able to build and deploy my Micronaut application to the ARM architecture on AWS Lambda.

Following the guide https://guides.micronaut.io/latest/mn-application-aws-lambda-graalvm-maven-java.html

$ ./mvnw package -Dpackaging=docker-native -Dmicronaut.runtime=lambda -Pgraalvm

gives me to following error on my MacBook Pro M1

```
[INFO] Removing intermediate container 14ec5ea17c9b
[INFO]  ---> ee5c7203c1a9
[INFO] Step 8/22 : RUN /usr/lib/graalvm/bin/gu install native-image
[INFO] 
[INFO]  ---> Running in 64fad8633c98
[INFO] /lib64/ld-linux-x86-64.so.2: No such file or directory

[ERROR] The command '/bin/sh -c /usr/lib/graalvm/bin/gu install native-image' returned a non-zero code: 255
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:22 min
[INFO] Finished at: 2021-12-03T14:16:20Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.micronaut.build:micronaut-maven-plugin:3.0.2:docker-native (default-docker-native) on project complete: Execution default-docker-native of goal io.micronaut.build:micronaut-maven-plugin:3.0.2:docker-native failed: Could not build image: The command '/bin/sh -c /usr/lib/graalvm/bin/gu install native-image' returned a non-zero code: 255 -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

I believe this is because https://github.com/micronaut-projects/micronaut-maven-plugin/blob/master/src/main/resources/dockerfiles/DockerfileAwsCustomRuntime specifically targets the x86 version of GraalVM.

--- a/src/main/java/io/micronaut/build/AbstractDockerMojo.java
+++ b/src/main/java/io/micronaut/build/AbstractDockerMojo.java
@@ -46,6 +46,8 @@ public abstract class AbstractDockerMojo extends AbstractMojo {
     public static final String LATEST_TAG = "latest";
     public static final String DEFAULT_BASE_IMAGE_GRAALVM_RUN = "frolvlad/alpine-glibc:alpine-3.12";
     public static final String MOSTLY_STATIC_NATIVE_IMAGE_GRAALVM_FLAG = "-H:+StaticExecutableWithDynamicLibC";
+    public static final String ARM_ARCH = "aarch64";
+    public static final String X86_64_ARCH = "amd64";
 
     protected final MavenProject mavenProject;
     protected final JibConfigurationService jibConfigurationService;
@@ -127,6 +129,18 @@ public abstract class AbstractDockerMojo extends AbstractMojo {
             graalVmJvmVersion = "java17";
         }
         return graalVmJvmVersion;
+    }
+
+    /**
+     * Detects the OS architecture to use for GraalVM depending on the <code>os.arch</code> system property.
+     */
+    protected String graalVmArch() {
+        String osArch = System.getProperty("os.arch");
+        if (ARM_ARCH.equals(osArch)) {
+            return ARM_ARCH;
+        } else {
+            return X86_64_ARCH;
+        }
     }
 
     /**

--- a/src/main/java/io/micronaut/build/DockerNativeMojo.java
+++ b/src/main/java/io/micronaut/build/DockerNativeMojo.java
@@ -118,10 +118,12 @@ public class DockerNativeMojo extends AbstractDockerMojo {
     private void buildDockerNativeLambda() throws IOException {
         BuildImageCmd buildImageCmd = dockerService.buildImageCmd(DockerfileMojo.DOCKERFILE_AWS_CUSTOM_RUNTIME)
                 .withBuildArg("GRAALVM_VERSION", graalVmVersion())
-                .withBuildArg("GRAALVM_JVM_VERSION", graalVmJvmVersion());
+                .withBuildArg("GRAALVM_JVM_VERSION", graalVmJvmVersion())
+                .withBuildArg("GRAALVM_ARCH", graalVmArch());
 
         getLog().info("Using GRAALVM_VERSION: " + graalVmVersion());
         getLog().info("Using GRAALVM_JVM_VERSION: " + graalVmJvmVersion());
+        getLog().info("Using GRAALVM_ARCH: " + graalVmArch());
 
         // Starter sets the right class in pom.xml:
         //   - For applications: io.micronaut.function.aws.runtime.MicronautLambdaRuntime

--- a/src/main/java/io/micronaut/build/DockerfileMojo.java
+++ b/src/main/java/io/micronaut/build/DockerfileMojo.java
@@ -173,6 +173,7 @@ public class DockerfileMojo extends AbstractDockerMojo {
                         result.add(line
                                 .replace("${GRAALVM_VERSION}", graalVmVersion())
                                 .replace("${GRAALVM_JVM_VERSION}", graalVmJvmVersion())
+                                .replace("${GRAALVM_ARCH}", graalVmArch())
                                 .replace("${GRAALVM_ARGS} ", graalVmBuildArgs)
                                 .replace("${CLASS_NAME}", mainClass)
                         );

--- a/src/main/resources/dockerfiles/DockerfileAwsCustomRuntime
+++ b/src/main/resources/dockerfiles/DockerfileAwsCustomRuntime
@@ -1,10 +1,10 @@
 FROM amazonlinux:2 AS graalvm
+ENV LANG=en_US.UTF-8
 ARG GRAALVM_VERSION
 ARG GRAALVM_JVM_VERSION
-ARG CLASS_NAME
-ENV LANG=en_US.UTF-8
+ARG GRAALVM_ARCH
 RUN yum update -y && yum install -y gcc gcc-c++ zlib-devel zip tar gzip && yum clean all
-RUN curl -4 -L https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-${GRAALVM_VERSION}/graalvm-ce-${GRAALVM_JVM_VERSION}-linux-amd64-${GRAALVM_VERSION}.tar.gz -o /tmp/graalvm.tar.gz \
+RUN curl -4 -L https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-${GRAALVM_VERSION}/graalvm-ce-${GRAALVM_JVM_VERSION}-linux-${GRAALVM_ARCH}-${GRAALVM_VERSION}.tar.gz -o /tmp/graalvm.tar.gz \
     && tar -zxf /tmp/graalvm.tar.gz -C /tmp \
     && mv /tmp/graalvm-ce-${GRAALVM_JVM_VERSION}-${GRAALVM_VERSION} /usr/lib/graalvm \
     && rm -rf /tmp/*
@@ -14,6 +14,7 @@ WORKDIR /home/app
 COPY classes /home/app/classes
 COPY dependency/* /home/app/libs/
 ARG GRAALVM_ARGS=""
+ARG CLASS_NAME
 RUN native-image ${GRAALVM_ARGS} -H:Class=${CLASS_NAME} -H:Name=application --no-fallback -cp "/home/app/libs/*:/home/app/classes/"
 
 FROM amazonlinux:2

--- a/src/site/asciidoc/examples/package.adoc
+++ b/src/site/asciidoc/examples/package.adoc
@@ -212,7 +212,11 @@ of the build steps expect a certain base image. By default, the native images ar
 
 In the case of AWS custom runtime, it starts from `amazonlinux:2`, and this cannot be changed. Also, in this case the
 result is not a tagged Docker image, but a `function.zip` archive that contains the launch script and the native binary.
-Essentially, what you need to upload to AWS Lambda.
+Essentially, what you need to upload to AWS Lambda. Also in this case, the Micronaut Maven Plugin will detect the host
+operating system architecture (based on the `os.arch` Java system property) and will install the corresponding GraalVM
+binary distribution inside the Docker image. This means that when running packaging from an X86_64 (Intel/AMD) machine,
+the produced native image will be an `amd64` binary, whilst on an ARM host (such as the new Mac M1) it will be an
+`aarch64` binary.
 
 You can pass additional arguments to the executable in the following way:
 

